lib/Sema/TypeCheckAttr.cpp

```diff
@@ -1876,17 +1876,26 @@ synthesizeMainBody(AbstractFunctionDecl *fn, void *arg) {
     // Pass main into _runAsyncMain(_ asyncFunc: () async throws -> ())
     // Resulting $main looks like:
     // $main() { _runAsyncMain(main) }
+    auto makeErrorExpr = [&context, mainFunction]() {
+      auto result = new (context) ErrorExpr(mainFunction->getSourceRange());
+      SmallVector<ASTNode, 1> stmts;
+      stmts.push_back(result);
+      auto body = BraceStmt::create(context, SourceLoc(), stmts, SourceLoc(),
+                                    /*Implicit*/ true);
+      return std::make_pair(body, /*typechecked*/ true);
+    };
     auto *concurrencyModule = context.getLoadedModule(context.Id_Concurrency);
     if (!concurrencyModule) {
       context.Diags.diagnose(mainFunction->getAsyncLoc(),
                              diag::async_main_no_concurrency);
-      auto result = new (context) ErrorExpr(mainFunction->getSourceRange());
-      SmallVector<ASTNode, 1> stmts;
-      stmts.push_back(result);
-      auto body = BraceStmt::create(context, SourceLoc(), stmts,
-                                    SourceLoc(), /*Implicit*/true);
+      return makeErrorExpr();
+    }
 
-      return std::make_pair(body, /*typechecked*/true);
+    if (!context.LangOpts.EnableExperimentalAsyncMain) {
+      context.Diags.diagnose(
+          mainFunction->getAsyncLoc(), diag::pound_error,
+          "pass '-enable-experimental-async-main' to use async-main");
```

Comment:
I think this error message's content should be revised for two reasons:
1. Users normally call `swiftc` so they would need `-Xfrontend` too.
2. We should probably say "'async' main function" instead of "async-main".

Perhaps we should follow the existing style of message, in `DiagnosticsParse.def` for `attr_requires_concurrency`, which says: "_ is only valid when experimental concurrency is enabled". Since this error would only come up once per project, it shouldn't be too noisy to follow up the error with a note that tells the user specifically what flag to use for async main.

Reply:
Oh and also, we should have a single regression test to make sure it's actually disallowed without the flag.